### PR TITLE
TextArea

### PR DESCRIPTION
**Pull request by Konstantin8105**
Create for #64
* Adding/removing characters
* Navigating the text
* Graphemes support
* Test for finding corner cases
* Add `demos` source
* Minimal change into `TextView`
* No color
* No regions
* No `Form` interface

### CODE REVIEW

File: textarea.go [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+				f.cursorByBuffer(line, pos)
+			}()
+		case tcell.KeyBackspace, tcell.KeyBackspace2:
+			line, pos = f.deleteRune()
+		default:
+			r := event.Rune()
+			if unicode.IsLetter(r) ||
+				unicode.IsNumber(r) ||
+				unicode.IsMark(r) ||
+				unicode.IsSpace(r) {
```
Comment (berrange) [inline textarea.go:421]: This filtering is too strict - it blocks user from entering any special characters, eg !"£$%^&*(){}~@:<>?  and more. I think it could be reduced to just  "unicode.IsGraphic", so that it only blocks control sequences.

**Comment by jpeletier** [comment]
@Konstantin8105 Given the pace at which reviews go in this repo, I'd suggest you move this widget to your own staond-alone repo under your control so we can begin using this and you can get PRs for improvements. I did this with my widget, [winman](https://github.com/epiclabs-io/winman) so it would be available immediately to everybody.

File: textarea.go (hunk 1 above)
Comment (Konstantin8105) [reply textarea.go:421]: I agree with you.

**Comment by Konstantin8105** [comment]
@jpeletier  In according to LICENSE, you can merge that widget in your project without any link on myself. Feel free to coping, using, and ...

**Comment by jpeletier** [comment]
@Konstantin8105 Yes, thank you for a permisive license, but I would like to avoid copying your code if I can better link to it and everyone shares the same to improve it together.

This repo is quite "slow". If you keep this in your repo, we can probably evolve your TextArea widget faster together.

**Comment by xxwassyxx** [comment]
This is actually a very very useful addition. Is there any reason it cannot be merged into master. Just trying to get a picture on timeline as this is also what I need.

**Comment by buzzdan** [comment]
watching this one as well...

**Comment by skanehira** [comment]
This is very useful!
I waiting and I hope TextArea added

**Comment by l13t** [comment]
Also waiting for merge of this PR.

**Comment by brlbil** [comment]
I have been waiting for this future to start a personal project. So I appreciate it if it can be merged asap.

**Comment by rivo** [comment]
I know this will be disappointing to some of you but it's unlikely I will merge this PR. There are a whole bunch of reasons but the two major reasons are: I don't think `TextArea` should be based on `TextView` because it would inherit all the shortcomings of `TextView`, mainly its inability to handle large texts as well as its non-optimal performance. The second major reason is that this PR has a bug at its core that will not be trivial to fix: Runes are not Unicode graphemes. A grapheme can consist of multiple runes. (A lot of emojis, e.g. flags, consist of multiple runes.) This will lead to bugs; I've already seen and fixed many related bugs.

Please fill out #578, `TextArea` is one of the features you can "vote" for. (So far, no one has voted for the text editing component. But there's still time to fill out the questionnaire.) If it comes out on top, it will be one of the next big things I will tackle.

File: textarea.go [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+	f.cursor.y = indexLine + y - f.view.lineOffset
+	if isOutsideBuffer {
+		f.cursor.x++
+	}
+	if y+height <= f.cursor.y {
+		diff := (indexLine - lastIndexLine) - (y + height - lastCy) + 1
+		f.view.lineOffset += diff
+	}
+	if f.cursor.y < y {
+		diff := -((lastIndexLine - indexLine) - lastCy) - 1
```
Comment (berrange) [inline textarea.go:352]: As best I can tell 'lastCy' shouldn't be involved in this calculation at all.  When the textarea fills the screen lastCy will be zero when we hit the top of the screen and need to scroll earlier lines into view.  When the textarea is offset from the top of the screen though, lastCy will be non-zero and this will cause `diff` to get ridiculous values. As a result lineOffset gets set to incorrect values and all your text disappears off the textarea display. Removing 'lastCy' from this calc made scrolling work as expected.

**Comment by rivo** [comment]
`TextArea` exists now. (Since last year already.)
